utility/src/main/java/org/oppia/android/util/math/MathTokenizer.kt

```diff
@@ -0,0 +1,441 @@
+import org.oppia.android.app.model.MathUnaryOperation.Operator.POSITIVE
+import org.oppia.android.util.math.PeekableIterator.Companion.toPeekableIterator
+import java.lang.StringBuilder
+
+/**
+ * Input tokenizer for math (numeric & algebraic) expressions and equations.
+ *
+ * See https://docs.google.com/document/d/1JMpbjqRqdEpye67HvDoqBo_rtScY9oEaB7SwKBBspss/edit for the
+ * grammar specification supported by this tokenizer.
+ *
+ * This class implements an LL(1) single-pass tokenizer with no caching. Use [tokenize] to produce a
+ * sequence of [Token]s from the given input stream.
+ */
+class MathTokenizer private constructor() {
+  companion object {
+    /**
+     * Returns a [Sequence] of [Token]s for the specified input string.
+     *
+     * Note that this tokenizer will attempt to recover if an invalid token is encountered (i.e.
+     * tokenization will continue). Further, tokenization occurs lazily (i.e. as the sequence is
+     * traversed), so calling this method is essentially zero-cost until tokens are actually needed.
+     * The sequence should be converted to a [List] if they need to be retained after initial
+     * tokenization since the sequence retains no memory.
+     */
+    fun tokenize(input: String): Sequence<Token> = tokenize(input.toCharArray().asSequence())
+
+    private fun tokenize(input: Sequence<Char>): Sequence<Token> {
+      val chars = input.toPeekableIterator()
+      return generateSequence {
+        // Consume any whitespace that might precede a valid token.
+        chars.consumeWhitespace()
+
+        // Parse the next token from the underlying sequence.
+        when (chars.peek()) {
+          in '0'..'9' -> tokenizeIntegerOrRealNumber(chars)
+          in 'a'..'z', in 'A'..'Z' -> tokenizeVariableOrFunctionName(chars)
+          '√' -> tokenizeSymbol(chars) { startIndex, endIndex ->
+            Token.SquareRootSymbol(startIndex, endIndex)
+          }
+          '+' -> tokenizeSymbol(chars) { startIndex, endIndex ->
+            Token.PlusSymbol(startIndex, endIndex)
+          }
+          '-', '−' -> tokenizeSymbol(chars) { startIndex, endIndex ->
+            Token.MinusSymbol(startIndex, endIndex)
+          }
+          '*', '×' -> tokenizeSymbol(chars) { startIndex, endIndex ->
+            Token.MultiplySymbol(startIndex, endIndex)
+          }
+          '/', '÷' -> tokenizeSymbol(chars) { startIndex, endIndex ->
+            Token.DivideSymbol(startIndex, endIndex)
+          }
+          '^' -> tokenizeSymbol(chars) { startIndex, endIndex ->
+            Token.ExponentiationSymbol(startIndex, endIndex)
+          }
+          '=' -> tokenizeSymbol(chars) { startIndex, endIndex ->
+            Token.EqualsSymbol(startIndex, endIndex)
+          }
+          '(' -> tokenizeSymbol(chars) { startIndex, endIndex ->
+            Token.LeftParenthesisSymbol(startIndex, endIndex)
+          }
+          ')' -> tokenizeSymbol(chars) { startIndex, endIndex ->
```

Comment:
Should we create some static fields for these variables? '=', ')', etc....

Reply:
I personally find it more readable to see the actual symbols being parsed, but it might be worth cross referencing for someone else's thoughts as far as readable is concerned if we differ on this point.

Beyond that, do you see any specific advantage to using constants?

Reply:
Not exactly, this readability point sounds good. Let's keep it unchanged.